structuralcodes/codes/mc2010/_concrete_material_properties.py

```diff
@@ -81,3 +84,192 @@ def Gf(fck: float) -> float:
+    """
+    _check_cem_strength_class(cem_class)
+    _check_agg_types(agg_type)
+    if cem_class.upper() in ["32.5 R", "42.5 N"]:
+        if _fcm > 60:
+            S = 0.20
+        else:
+            S = 0.25
+    elif cem_class.upper() in ["42.5 R", "52.5 N", "52.5 R"]:
+        S = 0.20
+    elif cem_class.upper() in ["32.5 N"]:
+        if _fcm > 60:
+            S = 0.20
+        else:
+            S = 0.38
+    if agg_type.lower() == 'basalt':
+        ALPHA_E = 1.2
+    elif agg_type.lower() == 'quartzite':
+        ALPHA_E = 1.0
+    elif agg_type.lower() == 'limestone':
+        ALPHA_E = 0.9
+    elif agg_type.lower() == 'sandstone':
+        ALPHA_E = 0.7
+
+    return {'S': S, 'alpha_e': ALPHA_E}
+
+
+def _calc_E28(
+    fc: float, TABULAR_VALUES: dict, fc_value_type: str = "char"
+) -> float:
+    """Calculate the modulus of elasticity for normal weight concrete
+        at 28 days.
+
+    Defined in fib Model Code 2010 (2013), Eq. 5.1-21.
+
+    args:
+        fc (float): The mean or characteristic value of the compressive
+            strength of the concrete in MPa.
+        TABULAR_VALUES (dict): Material constants for different cement
+            types and aggregate types as given by the fib Model Code
+            2010.
+        fc_value_type (str): The specified value of fc is "mean" or "char"
+            (default).
+
+    returns:
+        float: The modulus of elasticity for normal weight concrete
+            at 28 days in MPa.
+    """
+    if fc_value_type == "char":
+        _fcm = fc + 8
+    elif fc_value_type == "mean":
+        _fcm = fc
+    else:
+        raise ValueError(
+            "The value_type of fc can only be one of the "
+            "following: 'mean' or 'char'."
+        )
+
+    return _EC0 * TABULAR_VALUES['alpha_e'] * (_fcm / 10) ** (1 / 3)
+
+
+def _calc_beta_cc(time: float, TABULAR_VALUES: dict) -> float:
+    """Calculate multiplication factor beta_cc, used to determine the
+        modulus of elasticity at an arbitrary time.
+
+    Defined in fib Model Code 2010 (2013), Eq. 5.1-51.
+
+    args:
+        time (float): The time in days at which the modulus of
+            elasticity is to be determined.
+        TABULAR_VALUES (dict): Material constants for different cement
+            types and aggregate types as given by the fib Model Code
+            2010.
+
+    returns:
+        float: Multiplication factor beta_cc.
+    """
+    return np.exp(TABULAR_VALUES['S'] * (1 - np.sqrt(28 / time)))
+
+
+def _calc_beta_e(beta_cc: float) -> float:
+    """Calculate multiplication factor beta_e, used to determine the
+        modulus of elasticity at an arbitrary time.
+
+    Defined in fib Model Code 2010 (2013), Eq. 5.1-57.
+
+    args:
+        beta_cc (float): multiplication factor as defined in the fib Model
+            Code 2010 (2013), Eq. 5.1-51.
+
+    returns:
+        float: Multiplication factor beta_e.
+    """
+    return np.sqrt(beta_cc)
```

Comment:
Good point, but in the case of establishing a creep compliance matrix it would be useful to be able to work on arrays instead of element by element. `np.sqrt` accepts `ArrayLike`, so this could be a step in that direction 😃